crates/daphne/src/pine/flp.rs

```diff
@@ -174,27 +174,54 @@ impl<F: FftFriendlyFieldElement> PineType<F> {
         gradient: &[F],
         wr_joint_rand_seed: &Seed<16>,
     ) -> [F; NUM_WR_TESTS] {
+        debug_assert_eq!(gradient.len(), self.cfg.dimension);
         let mut xof = XofTurboShake128::seed_stream(
             wr_joint_rand_seed,
             &self.cfg.dst(USAGE_WR_JOINT_RAND),
             &[],
         );
-        let mut buf = vec![0_u8; chunk_count(4, NUM_WR_TESTS * self.cfg.dimension)];
-        xof.fill(&mut buf[..]);
+        let rand_len_per_test = chunk_count(4, self.cfg.dimension);
+        let mut rand = vec![0_u8; rand_len_per_test * NUM_WR_TESTS];
+        xof.fill(&mut rand[..]);
 
         let mut wr_test_results = [F::zero(); NUM_WR_TESTS];
-        let mut i = 0;
-        for wr_test_result in &mut wr_test_results {
-            for x in gradient {
-                // TODO spec: Consider reversing the order in which we read the byte. We can save a
-                // little computation if we can get rid of the subtraction below.
-                let rand_bits = (buf[i >> 3] >> (6 - (i & 7))) & 0b11;
-                match rand_bits {
-                    0b00 => *wr_test_result -= *x,
-                    0b11 => *wr_test_result += *x,
+        for (wr_test_result, rand_per_test) in wr_test_results
+            .iter_mut()
+            .zip(rand.chunks(rand_len_per_test))
+        {
+            for (gradient_chunk, r) in gradient.chunks(4).zip(rand_per_test) {
+                match r & 3 /* 0b11 */ {
+                    0 => *wr_test_result -= gradient_chunk[0],
+                    3 => *wr_test_result += gradient_chunk[0],
+                    _ => (),
+                };
```

Comment:
you can try this formula instead:

```rs
let n = (r & 0b11) //  n \in {0,1,2,3}
let a = gradient_chunk[0]
wr_test_result += a * ((n-1)>>1);
or
wr_test_result += a & ((n-1)>>1);

```

The match sentence can do a lot of comparisons. Verify whether this has some impact.

Reply:
I see what you're getting at, but I don't see how it works:
1. For the math two work, you need to convert `n` into the field, so that `-1` becomes `prime_modulus - 1`.
2. We can't do bit twiddling on the montgomery representation as far as I know

So what we want is something like (sorry about how verbose this is, we need to improve the field API in libprio):
```rust
a * (F::from(F::Integer::try_from(n as usize).unwrap()) - F::one()) * F::from(F::Integer::try_from(2).unwrap()).inv()
```

This costs two field multiplications per test and gradient element. Wouldn't this be more expensive than missing a branch prediction? 

Perhaps we can do the arithmetic in the integer representation of the field elements?

Reply:
Since the values are field elements, the bit tricks have to be expanded to field operations, which may result in overheads. 
Let's use continue using the current approach.